### PR TITLE
Compute System.Private.CoreLib ILLink roots required by the runtime.

### PR DESCRIPTION
**Pull request by erozenfeld**
Embed the result as a System.Private.CoreLib.dll xml resource
for consumption by ILLink.

Update mscorlib.h to make several dependencies explicit.

The actual msbuild task lives in the linker repo:
https://github.com/mono/linker/blob/master/corebuild/integration/ILLink.Tasks/CreateRuntimeRootDescriptorFile.cs

### CODE REVIEW

**Comment by erozenfeld** [comment]
@sbomer @weshaggard @jkotas PTAL /cc @JosephTremoulet

**Comment by jkotas** [comment]
What is the size of the embedded .xml file?

**Comment by erozenfeld** [comment]
43K

**Comment by erozenfeld** [comment]
Here is an example: https://gist.github.com/erozenfeld/b005d315c2f8e7da791bd40a125c07d9

**Comment by jkotas** [comment]
Sounds reasonable. The change looks good to me once CI is all green.

File: src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    <_MscorlibFilePath Condition=" '$(_MscorlibFilePath)' == '' ">$(MSBuildThisFileDirectory)..\vm\mscorlib.h</_MscorlibFilePath>
+    <_CortypeFilePath Condition=" '$(_CortypeFilePath)' == '' ">$(MSBuildThisFileDirectory)..\inc\cortypeinfo.h</_CortypeFilePath>
+    <_RexcepFilePath Condition=" '$(_RexcepFilePath)' == '' ">$(MSBuildThisFileDirectory)..\vm\rexcep.h</_RexcepFilePath>
+  </PropertyGroup>
+
+  <!-- Generates the xml root descriptor file for ILLink. The file contains roots required by the runtime. -->
+  <UsingTask TaskName="CreateRuntimeRootILLinkDescriptorFile" AssemblyFile="$(MSBuildThisFileDirectory)..\..\packages\illink.tasks\$(ILLinkTasksPackageVersion)\tools\net46\ILLink.Tasks.dll" />
+  <Target Name="_CreateILLinkRuntimeRootDescriptorFile"
+          Inputs="$(_NamespaceFilePath);$(_MscorlibFilePath);"
+          Outputs="$(_ILLinkRuntimeRootDescriptorFilePath);$(_CortypeFilePath);$(_RexcepFilePath)">
```
Comment (sbomer) [inline src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets]: Should _CortypeFilePath and _RexcepFilePath be inputs?

Comment (erozenfeld) [reply src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets]: Yes, they should be. Updated.

File: init-tools.cmd [hunk 2]
```diff
@@ -8,6 +8,7 @@ set DOTNET_PATH=%TOOLRUNTIME_DIR%\dotnetcli\
 if [%DOTNET_CMD%]==[] set DOTNET_CMD=%DOTNET_PATH%dotnet.exe
 if [%BUILDTOOLS_SOURCE%]==[] set BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
 set /P BUILDTOOLS_VERSION=< "%~dp0BuildToolsVersion.txt"
+if [%ILLINK_SOURCE%]==[] set ILLINK_SOURCE=https://dotnet.myget.org/F/dotnet-core/api/v3/index.json
```
Comment (weshaggard) [inline init-tools.cmd]: Lets try to avoid this. Can we instead put this package on the dotnet-buildtools feed?

File: init-tools.sh [hunk 3]
```diff
@@ -9,6 +9,7 @@ __DOTNET_CMD=$__DOTNET_PATH/dotnet
 if [ -z "$__BUILDTOOLS_SOURCE" ]; then __BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json; fi
 export __BUILDTOOLS_USE_CSPROJ=true
 __BUILD_TOOLS_PACKAGE_VERSION=$(cat $__scriptpath/BuildToolsVersion.txt)
+if [ -z "$__ILLINK_SOURCE" ]; then __ILLINK_SOURCE=https://dotnet.myget.org/F/dotnet-core/api/v3/index.json; fi
```
Comment (weshaggard) [inline init-tools.sh]: Same for this script we should avoid adding more feeds to restore the tools. All the dependencies should come from dotnet-buildtools feed.

**Comment by weshaggard** [comment]
Other then the putting the package on the buildtools feed the change looks reasonable.

**Comment by erozenfeld** [comment]
@sbomer What is the reason we are publishing the linker on dotnet-core rather than dotnet-buildtools?

**Comment by weshaggard** [comment]
It is OK to got to dotnet-core feed as well but for anything we want to consume it our toolset should be also put on the dotnet-buildtools feed.

**Comment by erozenfeld** [comment]
@sboemer Can you please look into putting the linker package to dotnet-buildtools in addition to dotnet-core?

**Comment by erozenfeld** [comment]
@dotnet-bot test CentOS7.1 x64 Checked Innerloop Build and Test 
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test
@dotnet-bot test Windows_NT arm Cross Checked Innerloop Build and Test 
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test
@dotnet-bot test Windows_NT armlb Cross Checked Innerloop Build and Test

**Comment by erozenfeld** [comment]
@weshaggard I changed this to pull the linker package from dotnet-buildtools feed. PTAL.
/cc: @JosephTremoulet @sbomer

File: src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    <_ILLinkRuntimeRootDescriptorFilePath Condition=" '$(_ILLinkRuntimeRootDescriptorFilePath)' == '' ">$(IntermediateOutputPath)$(_ILLinkRuntimeRootDescriptorFileName)</_ILLinkRuntimeRootDescriptorFilePath>
+    <_NamespaceFilePath Condition=" '$(_NamespaceFilePath)' == '' ">$(MSBuildThisFileDirectory)..\vm\namespace.h</_NamespaceFilePath>
+    <_MscorlibFilePath Condition=" '$(_MscorlibFilePath)' == '' ">$(MSBuildThisFileDirectory)..\vm\mscorlib.h</_MscorlibFilePath>
+    <_CortypeFilePath Condition=" '$(_CortypeFilePath)' == '' ">$(MSBuildThisFileDirectory)..\inc\cortypeinfo.h</_CortypeFilePath>
+    <_RexcepFilePath Condition=" '$(_RexcepFilePath)' == '' ">$(MSBuildThisFileDirectory)..\vm\rexcep.h</_RexcepFilePath>
+  </PropertyGroup>
+
+  <!-- Generates the xml root descriptor file for ILLink. The file contains roots required by the runtime. -->
+  <UsingTask TaskName="CreateRuntimeRootILLinkDescriptorFile" AssemblyFile="$(MSBuildThisFileDirectory)..\..\packages\illink.tasks\$(ILLinkTasksPackageVersion)\tools\net46\ILLink.Tasks.dll" />
```
Comment (weshaggard) [inline src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets]: `$(MSBuildThisFileDirectory)..\..\packages\illink.tasks\$(ILLinkTasksPackageVersion)\tools\net46\ILLink.Tasks.dll` that is making a number of bad assumptions.

1. It is assuming the packages directory. Use the PackagesDir property instead.
2. It is assuming we only run on windows on .NET Framework.  We need to be sure this works on .NET Core as well as on non-windows. See https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks.Feed/PackageFiles/Microsoft.DotNet.Build.Tasks.Feed.targets#L42 as an example of how other tasks work.

**Comment by weshaggard** [comment]
>@weshaggard I changed this to pull the linker package from dotnet-buildtools feed

Thanks that part looks good. I added some other feedback to your targets file though.

File: src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets (hunk 4 above)
Comment (erozenfeld) [reply src/mscorlib/CreateRuntimeRootILLinkDescriptorFile.targets]: @weshaggard I fixed the issues you pointed out. PTAL.

**Comment by erozenfeld** [comment]
@dotnet-bot test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by erozenfeld** [comment]
@dotnet-bot test Tizen armel Cross Checked Build
@dotnet-bot test Windows_NT arm64 Cross Checked normal Build and Test

**Comment by erozenfeld** [comment]
@dotnet-bot test CentOS7.1 x64 Release Innerloop Build and Test 
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test.*

**Comment by erozenfeld** [comment]
@dotnet-bot test Tizen armel Cross Checked Build

**Comment by erozenfeld** [comment]
@dotnet-bot test CentOS7.1 x64 Release Innerloop Build and Test

**Comment by erozenfeld** [comment]
@dotnet-bot test Tizen armel Cross Checked Build
@dotnet-bot test CentOS7.1 x64 Release Innerloop Build and Test

**Comment by erozenfeld** [comment]
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

**Comment by erozenfeld** [comment]
Ignoring CentOS7.1 x64 Release Innerloop Build and Test failure since it's failing in many other PRs.

**Pull request merged**